parser-typechecker/src/Unison/Share/Types.hs

```diff
@@ -0,0 +1,144 @@
+-- >>> codeserverFromURI . fromJust $ parseURI "https://share.unison-lang.org/api"
+-- Just https://share.unison-lang.org:443/api
+-- >>> codeserverFromURI . fromJust $ parseURI "http://share.unison-lang.org/api"
+-- Just http://share.unison-lang.org:80/api
+codeserverFromURI :: URI -> Maybe CodeserverURI
+codeserverFromURI URI {..} = do
+  URIAuth {uriUserInfo, uriRegName, uriPort} <- uriAuthority
+  scheme <- case uriScheme of
+    "http:" -> Just Http
+    "https:" -> Just Https
+    _ -> Nothing
+  port <- case uriPort of
+    "" -> case scheme of
+      Http -> Just 80
+      Https -> Just 443
+    (':' : p) -> readMaybe p
+    _ -> Nothing
+  pure $
+    CodeserverURI
+      { codeserverScheme = scheme,
+        codeserverUserInfo = uriUserInfo,
+        codeserverRegName = uriRegName,
+        codeserverPort = port,
+        codeserverPath =
+          let unprefixed =
+                case uriPath of
+                  ('/' : path) -> path
+                  path -> path
+           in case List.splitOn "/" unprefixed of
+                [""] -> []
+                p -> p
+      }
+
+-- | This is distinct from the codeserver URI in that we store credentials by a normalized ID, since it's
+-- much easier to look up that way than from an arbitrary path.
+-- We may wish to use explicitly named configurations in the future.
+-- This currently uses a stringified uriAuthority.
+newtype CodeserverId = CodeserverId {codeserverId :: Text}
+  deriving newtype (Show, Eq, Ord, ToJSON, FromJSON, ToJSONKey, FromJSONKey)
+
+-- | Gets the part of the CodeserverURI that we use for identifying that codeserver in
+-- credentials files.
+--
+-- >>> import Data.Maybe (fromJust)
+-- >>> import Network.URI (parseURI)
+-- >>> codeserverIdFromURI (CodeserverURI . fromJust $ parseURI "http://localhost:5424/api")
+-- >>> codeserverIdFromURI (CodeserverURI . fromJust $ parseURI "https://share.unison-lang.org/api")
+-- Right "localhost"
+-- Right "share.unison-lang.org"
+codeserverIdFromURI :: URI -> Either Text CodeserverId
+codeserverIdFromURI uri =
+  case uriAuthority uri of
+    Nothing -> Left $ "No URI Authority for URI " <> tShow uri
+    Just ua -> pure $ codeserverIdFromURIAuth ua
+
+-- | Builds a CodeserverId from a URIAuth
+codeserverIdFromURIAuth :: URIAuth -> CodeserverId
+codeserverIdFromURIAuth ua =
+  (CodeserverId (Text.pack $ uriUserInfo ua <> uriRegName ua <> uriPort ua))
+
+-- | Gets the CodeserverId for a given CodeserverURI
+codeserverIdFromCodeserverURI :: CodeserverURI -> CodeserverId
+codeserverIdFromCodeserverURI =
+  codeserverIdFromURIAuth . codeserverAuthority
+
+-- | Builds a servant-compatible BaseUrl for a given CodeserverURI.
+codeserverBaseURL :: CodeserverURI -> Servant.BaseUrl
+codeserverBaseURL (CodeserverURI {..}) =
+  let scheme = case codeserverScheme of
+        Https -> Servant.Https
+        Http -> Servant.Http
+      host = codeserverUserInfo <> codeserverRegName <> ":" <> show codeserverPort
```

Comment:
This is how it would appear when provided, but that's a good a point.

Moving it to BasicAuth would require re-writing pieces of the Authenticated Client middleware; so I vote we leave it like this until anyone actually needs it, e.g. for a private unison-cloud or something, but I think we'd prefer they didn't use BasicAuth for that 🙃